### DISCUSSION TITLE
How to build extensions with unreleased versions of their language server?

### DISCUSSION THREAD

**Discussion by ssbarnea** [microsoft/vscode-discussions#36] [Extension Development QnA] [answered]
I am facing some problems with use of npm when trying to include unreleased versions on our own language server with the extension, basically because `npm link` is got broken years ago and nobody fixed it since.

As part of our testing pipelines and even development we often want to also build and test with future release of the language server, which is a different repository/project.

The big problem is that if you link that npm package it gets overridden when you run other npm commands, even silently so you can sometimes endup not testing the code that you wanted to test with.

Is there a better alternative that allows us to easily test with multiple versions of a package (like the current release from npm and the code from its main branch).

I was wondering if `yarn` or `npmpm` may be better suited for these, or if another solution is available.

As 4/5 changes we do usually involve both, I was even wondering if it would not be better to switch to a monorepo-like setup and avoid having two different projects fro extension itself and the language server. That approach is a little bit tricky because our language-server aims to also be compatible with other editors and we already have others that implemented extensions for vim and emacs. We would still publish it to npm, so maybe is not a big dead. Another advantage of having a single repository is that we no longer have to bother moving issues between the two as most users have no clue where a specific bug originates from, as not having a clue about what a language server is (kinda expected IMHO).

Any ideas or examples of an extension project that addressed the described problem?

PS. The projects where I face this issue are https://github.com/ansible/vscode-ansible and https://github.com/ansible/ansible-language-server

**Comment by isidorn** [comment]
I do not know. Maybe @dbaeumer has a tip for this setup

**Comment by dbaeumer** [comment, marked as answer]
You might want to look at https://github.com/microsoft/vscode-languageserver-node/blob/8b56a58d0a706bc0cc8e947b42ad135faf488fe3/build/bin/symlink.js#L9. I use `shelljs` to setup my own links.
